test/cpp/tensorexpr/tutorial.cpp

```diff
@@ -0,0 +1,366 @@
+    // is to use so called expression handles (as opposed to raw expressions
+    // like we did in the previous example). Expression handles overload common
+    // operations and allow us to express the same semantics in a more natural
+    // way:
+    ExprHandle l = 1;
+    ExprHandle r = Var::make("x", kInt);
+    ExprHandle m = l * r;
+    std::cout << "Tensor expression: " << *m.node() << std::endl;
+    // Prints: Tensor expression: 1 * x
+
+    // In a similar fashion we could construct arbitrarily complex expressions
+    // using mathematical and logical operations, casts between various data
+    // types, and a bunch of intrinsics.
+    ExprHandle a = Var::make("a", kInt);
+    ExprHandle b = Var::make("b", kFloat);
+    ExprHandle c = Var::make("c", kFloat);
+    ExprHandle x = ExprHandle(5) * a + b / (sigmoid(c) - 3.0f);
+    std::cout << "Tensor expression: " << *x.node() << std::endl;
+    // Prints: Tensor expression: float(5 * a) + b / ((sigmoid(c)) - 3.f)
+
+    // An ultimate purpose of tensor expressions is to optimize tensor
+    // computations, and in order to represent accesses to tensors data, there
+    // is a special kind of expression - a load.
+    // To construct a load we need two pieces: the base and the indices. The
+    // base of a load is a Buf expression, which could be thought of as a
+    // placeholder similar to Var, but with dimensions info.
+    //
+    // Let's construct a simple load:
+    BufHandle A("A", {ExprHandle(64), ExprHandle(32)}, kInt);
+    ExprHandle i = Var::make("i", kInt), j = Var::make("j", kInt);
+    ExprHandle load = Load::make(A.dtype(), A, {i, j}, /* mask= */ 1);
+    std::cout << "Tensor expression: " << *load.node() << std::endl;
+    // Prints: Tensor expression: A[i, j]
+  }
+
+  std::cout << "*** Tensors, Functions, and Placeholders ***" << std::endl;
+  {
+    // A tensor computation is represented by objects of Tensor class and
+    // consists of the following pieces:
+    //   - domain, which is specified by a Buf expression
+    //   - an expression (or several expressions if we want to perform several
+    //   independent computations over the same domain) for its elements, as a
+    //   function of indices
+    //
+    // We use Function objects to represent this. Let's build one:
+    std::vector<const Expr*> dims = {new IntImm(64), new IntImm(32)};
+
+    // Function arguments:
+    const Var* i = new Var("i", kInt);
+    const Var* j = new Var("j", kInt);
+    std::vector<const Var*> args = {i, j};
+    // Element expressions:
+    Expr* func_body1 = new Mul(i, j);
+    Expr* func_body2 = new Add(i, j);
+
+    Function* func =
+        new Function({"X", "Y"}, dims, args, {func_body1, func_body2});
+    std::cout << "Tensor function: " << *func << std::endl;
+    // Prints:
+    // Tensor function: Function F(i[64], j[32]) {
+    //   X = i * j
+    //   Y = i + j
+    // }
+
+    // A Tensor refers to an individual computation defined by a Function. For
```

Comment:
does a Function always return a new Tensor? i guess i'm a still a little confused abt the difference between a Function and a Block

Reply:
> does a Function always return a new Tensor?

No, technically when we create a Function we don't create any Tensors. However, to create a Tensor one need to create a Function.

> i'm a still a little confused abt the difference between a Function and a Block
Function and Tensor are very close concepts, but Function and Block are very different. Function here is not a function in programming languages sense (which can be also called method/procedure in some context). Here Function is a function in a mathematical sense, i.e. it's a formula, something like f(x) = x^2.

Reply:
@ngimel and I also find this comment to be unclear.